### DISCUSSION TITLE
feat: Whatsapp campaigns

### DISCUSSION THREAD

**Discussion by muhsin-k** [chatwoot/chatwoot#8465] [Feature Requests]
### Problem

Chatwoot enables businesses to reach their customers through WhatsApp. Adding the ability to send pre-approved whats app templates to the set of contacts will help more businesses.

### How

We will continue over our existing one-off campaign feature and try to satisfy the requirement with minimal changes.

### Requirements

- As a user, I want to send a WhatsApp campaign to a group of contacts
- As a user, I want to schedule a campaign.

### Engineering Spec

**Model Changes**

- Add the `template_params` attribute to accommodate template variables.

 **Logic**

- Contact labels satisfy the requirement to group contacts.
- WhatsApp campaign after trigger shouldn't be editable.
- Create a new service for WhatsApp campaign execution. (Reference app/services/sms/oneoff_sms_campaign_service.rb)
- Create a new sidekiq cron job that runs every 5 minutes, checks for any un-triggered WhatsApp campaign whose execution time has passed, and will trigger it.

### Front-end changes

1. The WhastApp campaign will accommodate in one-off campaigns page.
2. Empty text will be **Please create an sms or WhatsApp inbox and start adding campaigns**.
3. When clicking on the "Create campaign button"
4. Show Title and Choose inbox.
5. Choose Whatsapp inbox → Available templates, Variable support, audience and scheduled time.
6. Choose SMS inbox → Message, audience and scheduled time.

### Design mocks

![CleanShot 2023-02-03 at 00 40 48](https://user-images.githubusercontent.com/12408980/216427148-91a0df78-2a04-44d6-b613-9bde890fd152.png)

### Related issues

1. https://github.com/chatwoot/chatwoot/issues/4201
2. https://github.com/chatwoot/chatwoot/issues/5361
3. https://github.com/chatwoot/chatwoot/issues/5851

**Comment by ninoppp** [comment]
Is this feature coming in v3?

**Comment by glauberssoutodossantos** [comment]
Hi folks.

Here in Brazil we development something for this issue :). In feel weeks we will share the PR.

![image](https://github.com/chatwoot/chatwoot/assets/99494469/fa2eb26e-a99d-49bc-834c-dd5a1d79e85c)

**Reply by devangn** [reply]
@glauberssoutodossantos  Have you find sometime to send the PR please?

**Comment by jochenstu** [comment]
Would be very interested to know, how dev on this is coming along

**Comment by iamshouvikmitra** [comment]
Any updates @KarzounApps ?

**Comment by PedroCanoasFlip** [comment]
> Hi folks.
> 
> Here in Brazil we development something for this issue :). In feel weeks we will share the PR.
> 
> ![image](https://user-images.githubusercontent.com/99494469/237388047-fa2eb26e-a99d-49bc-834c-dd5a1d79e85c.png)

Hello friend, I'm Brazilian and I'm working in the same area as well.
I'm having problems with the message trigger, can you share with us how you did it?
more especially in the part that takes the message template and does the text formatting and sending it.

**Comment by sojan-official** [comment]
> Hi folks.
> 
> Here in Brazil we development something for this issue :). In feel weeks we will share the PR.
> 
> ![image](https://user-images.githubusercontent.com/99494469/237388047-fa2eb26e-a99d-49bc-834c-dd5a1d79e85c.png)

@glauberssoutodossantos just checking in to see if you folks are still interested in raising a PR for this.

**Reply by rohit-yelmar** [reply]
@glauberssoutodossantos  Hey, I'm also working on the same issue. Would love if I you could guide me through the problem and also if you could share this with community.

**Comment by macmaina** [comment]
> Hi folks.
> 
> Here in Brazil we development something for this issue :). In feel weeks we will share the PR.
> 
> ![image](https://user-images.githubusercontent.com/99494469/237388047-fa2eb26e-a99d-49bc-834c-dd5a1d79e85c.png)

Hi @glauberssoutodossantos 

Is this feature now available in V3?

**Comment by giquieu** [comment]
@glauberssoutodossantos Are you still going to send the PR? Would appreciate if this can be shared with the community! :)

**Comment by eusden** [comment]
@glauberssoutodossantos would appreciate if this can be shared with the community!

**Comment by ashishkaura8** [comment]
Any update  on this?

**Comment by pranavrajs** [comment]
Moving these to discussions as we will not be prioritizing these immediately.

**Comment by Marlonms** [comment]
Hello all,
We'd love to include the WhatsApp message sending campaign.
@pranavrajs, is there any roadmap to add this functionality?
@glauberssoutodossantos, is this amazing development available even if it's not officially added? This way, other folks (as me 😅) who need to use it can include it in their Chatwoot instances while the project includes the feature

**Reply by luiszxcv** [reply]
+1, new updates?  would appreciate if this can be shared with the community!

**Comment by yuliassn** [comment]
This is such an interesting functionality. Is there any update on the development?

**Comment by anasaldib** [comment]
We need it so bad

**Comment by luiszxcv** [comment]
@glauberssoutodossantos novidades? Tenho interesse nessa funcionalidade! Conseguiu desenvolve-la? Me chame no Whats para falarmos +5566996181212

**Comment by luiszxcv** [comment]
Chatwoot team, do we have a forecast for this feature? I am using several different software to meet this need today.

**Comment by zou-yu** [comment]
Really need this feature! 🙏

**Comment by janik-sauerbier** [comment]
Amazing feature! Would love to see it in Chatwoot!

**Comment by jcmatoskx** [comment]
Since Chatwoot is a multichannel platform is would make more sense if these would be just "Campaigns" and not  " Whatsapp Campaigns" and you could select the channel you wish to launch the campaign on.
Ideally with something like how modular email automation works  :)

**Comment by luiszxcv** [comment]
News? Is very important!

**Reply by luiszxcv** [reply]
@pranavrajs ?

**Comment by gauravgill95** [comment]
@pranavrajs any updates on this

**Comment by guiz01** [comment]
@pranavrajs news?

**Comment by Abodolla** [comment]


**Reply by janik-sauerbier** [reply]
What's the news? :)

**Reply by Abodolla** [reply]
how to DM @janik-sauerbier

**Reply by janik-sauerbier** [reply]
@Abodolla [Redacted]

**Reply by rohit-yelmar** [reply]
@Abodolla Could you please share the insights regarding the feature. It would be much appreciated if you could guide me implementing it.

**Reply by frycrosoft** [reply]
and for me and for me!! )) thx a lot

**Comment by FacundoSuarez-46** [comment]


**Reply by jochenstu** [reply]
How about you send a pull request instead? I don't think it's good form to advertise your company in a GH issue of an open source project.

**Reply by FacundoSuarez-46** [reply]
> How about you send a pull request instead? I don't think it's good form to advertise your company in a GH issue of an open source project.

Impossible to do that, i have an agreement with my client so that's not possible

**Comment by sarahmanning190** [comment]
@sojan-official Is this feature coming into v4?

**Comment by mudora23** [comment]
What are our options right now? I'd like to use a single number to send WhatsApp campaigns while also using Chatwoot for customer service.

**Comment by BT-Rajan** [comment]
@sojan-official @pranavrajs  do we have this in our roadmap or this feature is dropped ?

**Comment by badereddineqodia** [comment]
+1

**Comment by javierg** [comment]
👍

**Comment by badereddineqodia** [comment]
We are currently implementing these changes and will open a pull request once we’ve applied all required best practices and completed comprehensive testing.

**1- 1. WhatsApp Template Library:**
![image](https://github.com/user-attachments/assets/1db1d3ec-7ec0-4844-b08a-1fbc9afb3bb5)

**2-Campaign Toggle with WhatsApp Inbox Support & Dynamic Templates:**
![image](https://github.com/user-attachments/assets/5d6fe970-e71c-4c70-983d-b7cfcfab9e91)

**3- Template Creation, Approval Workflow & History Tracking:**
![image](https://github.com/user-attachments/assets/da2e7729-8ad8-420a-857e-1a8ce3929d9d)

**Reply by Marlonms** [reply]
At last! Lifesaver! Thanks so much!

**Reply by sojan-official** [reply]
@badereddineqodia ❤️  would be happy to collaborate over a PR !!

**Comment by pauloappbr** [comment]
Algum atualização dessa nova funcionalidade? isso ajudaria muito no meu cenário, para não recorrer a softwares de terceiros.

**Comment by sojan-official** [comment]
This will be available in the 4.4 release . ref: https://github.com/chatwoot/chatwoot/pull/11910
